File: packages/platform/src/ipc/blacklist.ts
/**
 * 黑名单管理 IPC 封装 — 个人黑名单 CRUD + 平台黑名单查询。
 *
 * 后端：壳层 `commands/blacklist.rs`（InMemoryBlacklistStore + app::blacklist::BlacklistService）。
 *
 * @author agent
 * @created 2026-08-13
 */

import { call } from "./invoke";

/** 个人黑名单条目（与 Rust `PersonalBlacklistItem` 对齐）。 */
export interface PersonalBlacklistItem {
  id: number;
  owner_id: number;
  account_id: string | null;
  buyer_id: string;
  buyer_nick: string | null;
  item_id: string | null;
  reason: string | null;
  is_enabled: boolean;
  created_at: string | null;
}

/** 命中级别：商品级 > 账户级 > 用户级（与 Rust `level()` 对齐）。 */
export function blacklistLevel(item: PersonalBlacklistItem): string {
  if (item.account_id && item.item_id) return "商品级";
  if (item.account_id) return "账户级";
  return "用户级";
}

/** 平台黑名单条目。 */
export interface PlatformBlacklistItem {
  id: number;
  owner_id: number;
  buyer_id: string;
  buyer_nick: string | null;
  created_at: string | null;
}

/** 查询个人黑名单。 */
export function blacklistPersonalList(request: {
  owner_id: number;
  page: number;
  page_size: number;
  buyer_id?: string;
  buyer_nick?: string;
}): Promise<[PersonalBlacklistItem[], number]> {
  return call<[PersonalBlacklistItem[], number]>("blacklist_personal_list", {
    request: {
      owner_id: request.owner_id,
      page: request.page,
      page_size: request.page_size,
      buyer_id: request.buyer_id ?? "",
      buyer_nick: request.buyer_nick ?? "",
    },
  });
}

/** 查询平台黑名单。 */
export function blacklistPlatformList(ownerId: number): Promise<[PlatformBlacklistItem[], number]> {
  return call<[PlatformBlacklistItem[], number]>("blacklist_platform_list", { ownerId });
}

/** 新建个人黑名单（buyer_ids 支持逗号/换行分隔批量）。 */
export function blacklistPersonalCreate(request: {
  owner_id: number;
  buyer_ids: string;
  account_id?: string;
  item_id?: string;
  reason?: string;
}): Promise<PersonalBlacklistItem[]> {
  return call<PersonalBlacklistItem[]>("blacklist_personal_create", {
    request: {
      owner_id: request.owner_id,
      buyer_ids: request.buyer_ids,
      account_id: request.account_id,
      item_id: request.item_id,
      reason: request.reason,
    },
  });
}

/** 切换启用状态。 */
export function blacklistSetEnabled(ownerId: number, id: number, enabled: boolean): Promise<void> {
  return call<void>("blacklist_set_enabled", {
    request: { owner_id: ownerId, id, enabled },
  });
}

/** 删除。 */
export function blacklistDelete(ownerId: number, id: number): Promise<void> {
  return call<void>("blacklist_delete", { request: { owner_id: ownerId, id } });
}

File: apps/desktop/src-tauri/src/platforms/xianyu/ipc/blacklist.rs
//! 黑名单管理 Tauri commands。

use crate::platforms::xianyu::persist::InMemoryBlacklistStore;
use crate::shared::ipc::IpcResponse;
use app::blacklist::{
    BlacklistQuery, BlacklistService, PersonalBlacklistItem, PlatformBlacklistItem,
};
use common;
use serde::Deserialize;
use std::sync::Arc;
use tauri::State;

pub struct BlacklistHandle {
    pub store: Arc<InMemoryBlacklistStore>,
}

#[derive(Debug, Deserialize)]
pub struct PersonalBlacklistListRequest {
    pub owner_id: i64,
    pub page: u32,
    pub page_size: u32,
    #[serde(default)]
    pub buyer_id: String,
    #[serde(default)]
    pub buyer_nick: String,
}

#[derive(Debug, Deserialize)]
pub struct PersonalBlacklistCreateRequest {
    pub owner_id: i64,
    pub buyer_ids: String,
    pub account_id: Option<String>,
    pub item_id: Option<String>,
    pub reason: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct BlacklistEnabledRequest {
    pub owner_id: i64,
    pub id: i64,
    pub enabled: bool,
}

#[derive(Debug, Deserialize)]
pub struct BlacklistDeleteRequest {
    pub owner_id: i64,
    pub id: i64,
}

#[tauri::command]
pub fn blacklist_personal_list(
    state: State<'_, BlacklistHandle>,
    request: PersonalBlacklistListRequest,
) -> common::DingDaResult<IpcResponse<(Vec<PersonalBlacklistItem>, u32)>> {
    let service = BlacklistService::new(state.store.as_ref());
    let query = BlacklistQuery {
        page: request.page,
        page_size: request.page_size,
        buyer_id: request.buyer_id,
        buyer_nick: request.buyer_nick,
    };
    let result = service
        .list_personal(request.owner_id, &query)
        .map_err(common::DingDaError::wrap)?;
    Ok(IpcResponse::ok(result))
}

#[tauri::command]
pub fn blacklist_platform_list(
    state: State<'_, BlacklistHandle>,
    owner_id: i64,
) -> common::DingDaResult<IpcResponse<(Vec<PlatformBlacklistItem>, u32)>> {
    let service = BlacklistService::new(state.store.as_ref());
    let result = service
        .list_platform(owner_id, &BlacklistQuery::default())
        .map_err(common::DingDaError::wrap)?;
    Ok(IpcResponse::ok(result))
}

#[tauri::command]
pub fn blacklist_personal_create(
    state: State<'_, BlacklistHandle>,
    request: PersonalBlacklistCreateRequest,
) -> common::DingDaResult<IpcResponse<Vec<PersonalBlacklistItem>>> {
    let service = BlacklistService::new(state.store.as_ref());
    let ids: Vec<&str> = request
        .buyer_ids
        .split(',')
        .map(str::trim)
        .filter(|id| !id.is_empty())
        .collect();
    if ids.is_empty() {
        return Err("买家 ID 不能为空".into());
    }
    let mut created = Vec::with_capacity(ids.len());
    for buyer_id in ids {
        let item = service
            .create(
                request.owner_id,
                buyer_id,
                request.account_id.as_deref(),
                request.item_id.as_deref(),
                request.reason.as_deref(),
            )
            .map_err(common::DingDaError::wrap)?;
        created.push(item);
    }
    Ok(IpcResponse::ok(created))
}

#[tauri::command]
pub fn blacklist_set_enabled(
    state: State<'_, BlacklistHandle>,
    request: BlacklistEnabledRequest,
) -> common::DingDaResult<IpcResponse<()>> {
    let service = BlacklistService::new(state.store.as_ref());
    service
        .set_enabled(request.owner_id, request.id, request.enabled)
        .map_err(common::DingDaError::wrap)?;
    Ok(IpcResponse::ok(()))
}

#[tauri::command]
pub fn blacklist_delete(
    state: State<'_, BlacklistHandle>,
    request: BlacklistDeleteRequest,
) -> common::DingDaResult<IpcResponse<()>> {
    let service = BlacklistService::new(state.store.as_ref());
    service
        .delete(request.owner_id, request.id)
        .map_err(common::DingDaError::wrap)?;
    Ok(IpcResponse::ok(()))
}
